core/src/main/scala/caliban/validation/Validator.scala

```diff
@@ -248,16 +251,26 @@ object Validator {
       }
     loop(selectionSet)
     val directiveLists                            = builder.result()
-    ZIO.foreachDiscard(directiveLists)(list => checkDirectivesUniqueness(list.map(_._1))).as(directiveLists.flatten)
+    ZIO
+      .foreachDiscard(directiveLists)(list => checkDirectivesUniqueness(list.map(_._1), directiveDefintions))
+      .as(directiveLists.flatten)
   }
 
-  private def checkDirectivesUniqueness(directives: List[Directive]): IO[ValidationError, Unit] =
-    directives.groupBy(_.name).find { case (_, v) => v.length > 1 } match {
+  private def checkDirectivesUniqueness(
+    directives: List[Directive],
+    directiveDefinitions: Map[String, List[DirectiveDefinition]]
+  ): IO[ValidationError, Unit] =
+    directives
+      .groupBy(_.name)
+      .find { case (n, v) =>
+        // only non-repeatable directives count against uniqueness
+        !directiveDefinitions.get(n).exists(_.exists(_.repeatable)) & (v.length > 1)
```

Comment:
Shouldn't it be `&&`?

Reply:
I can make it `&&` if you like, There's no data dependency between the two nor much computation to be saved with a short circuit, and in such cases I have a habit of using `&`. I don't think it'll make a noticeable difference here either way but `&` can be friendlier to CPU instruction pipelines.

I suppose the main argument for `&&` is that `v` is a `List` and so `v.length` is not a constant time call, but my preferred solution for that is to not use `List`

Reply:
I forgot when it was added but wouldn't it be more performant to run the length check first using `lengthIs`?

```scala
(v.lengthIs > 1)
```

That should be effectively constant time since it only ever has to check if there are at least two items in the list. That also saves on having to do the `repeatability` check against all directives

Reply:
I would have changed it to `lengthIs` but that's not available in 2.12. I could use `v.lengthCompare(1) > 0`, which is a less clear version of `lengthIs > 1`

I would expect no performance difference when using `&` since it does not short circuit and both sides get evaluated

Using `lengthCompare` first coupled with `&&` would likely have the best behavior in pathological cases where many directives are repeated many times. In more usual cases I expect these lists would have very few elements and you'd probably need a microbenchmark to tell them apart

I'd probably choose `lengthCompare` with `&&` myself, but I'll go with whatever you want to read and maintain

Reply:
How about simply `(v.lengthIs > 1) && !directiveDefinitions.get(n).exists(_.exists(_.repeatable))` 😄 
(we're nitpicking, it's not gonna make much of a different anyway)